pkg/docker/util.go

```diff
@@ -344,7 +367,7 @@ func pullAndCheck(image string, docker Docker, pullPolicy api.PullPolicy, config
 		return nil, err
 	}
 
-	err = CheckAllowedUser(docker, image, config.AllowedUIDs, r.OnBuild)
+	err = CheckAllowedUser(docker, image, config.AllowedUIDs, r.OnBuild, config.AssembleUser)
```

Comment:
is AssembleUser set to the value of the assemble-user label from the image at this point, if one was present?

Reply:
not for the builder image. However, the logic in `extractAssembleUser` (added to CheckAllowedUser) inspects the `assemble-user` label on the image if the override is empty.

Reply:
ok.  I feel like there is a hierarchy/priority ordering here that could be managed in a more obvious way..right now it feels like bits of it are scattered throughout the code.

the hierarchy is:

1) if alloweduids isn't set, we don't care what the assemble user is
2) if the config/cli explicitly provided an assemble user, check that value only (break)
3) if the image provides an assemble user label, check that value only (break)
4) if it's not on-build/layeredbuild check the default user of the image (break)
5) if it's a layered build check the USER logic in the on-build image (break)

factoring all that logic out to a single location would make it less fragile and easier to follow i think.

Reply:
i should say i *think* that is the hierarchy.

Reply:
> Challenge is how to handle OnBuild - do we need to check the onbuilder users as well as the default image user?

As I recall, the on-build logic checks the "on-build" commands in the base image to see if any are "USER" statements, and if any are "USER" statements, then we need to check what user they specify to ensure it is an allowed uid.

the last "USER" statement in the on-build args would become the default user for the image it produces. If there are no "USER" statements, then the default user for the resulting image is going to be the existing default user for the builder image.  (ie it's the same scenario as a non layered build, effectively).

Reply:
so it seems our hierarchy needs be reversed a bit (and the logic is a bit more complex):

1. No allowed uids - return nil
2. If layered, check OnBuild users. If ok then **continue**
3. If config has AssembleUser, check & return
4. If image has assemble user label, check & return
5. Fall through - check default image user